compiler/rustc_data_structures/src/tagged_ptr/copy.rs

```diff
@@ -143,14 +143,15 @@ where
             // `{non_zero} | packed_tag` can't make the value zero.
 
             let packed = (addr.get() >> T::BITS) | packed_tag;
-            unsafe { NonZeroUsize::new_unchecked(packed) }
+            unsafe { NonZero::<usize>::new_unchecked(packed) }
```

Comment:
If part of the goal is to get experience with type inference, then most of these calls to `NonZero::new` and `NonZero::new_unchecked` should probably be attempted *without* turbofish. That is:
```suggestion
            unsafe { NonZero::new_unchecked(packed) }
```

Reply:
I agree, it would be more appealing to sequence this PR after a different PR that makes `NonZero::new` into a single generic function.

As is, `NonZero::new` is currently ambiguous.

```console
error[E0034]: multiple applicable items in scope
 --> src/main.rs:6:22
  |
6 |     let _ = NonZero::new(0usize);
  |                      ^^^ multiple `new` found
  |
  = note: candidate #1 is defined in an impl for the type `NonZero<u8>`
  = note: candidate #2 is defined in an impl for the type `NonZero<u16>`
  = note: candidate #3 is defined in an impl for the type `NonZero<u32>`
  = note: candidate #4 is defined in an impl for the type `NonZero<u64>`
  = note: and 8 others
```

Reply:
> I don't think there is a good way for `NonZero` to be generic (apart from the struct definition) _before_ stabilizing because the docs will be a mess otherwise.

@reitermarkus how bad are the docs if we only change `new` and maybe any other constructors?

Reply:
> As is, `NonZero::new` is currently ambiguous.

Ah, I didn't realize that type inference wouldn't try to figure out which impl to use when they're distinct.

Reply:
I mean, the docs will be pretty bad as soon as we have any generic method since they will be split between the unstable `NonZero` type and all the type-specific aliases.